[libc++] Fix strict aliasing violation for `deque::const_iterator`

frederick-vs-ja opened the pull request:
When the allocators use fancy pointers, the internal map of `deque` stores `FancyPtr<T>` objects, and the previous strategy accessed these objects via `const FancyPtr<const T>` lvalues, which usually caused core language undefined behavior. Now `const_iterator` stores `FancyPtr<const FancyPtr<T>>` instead of `FancyPtr<const FancyPtr<const T>>`, and ABI break can happen when such two types have incompatible layouts.

This is necessary for reducing undefined behavior and `constexpr` support for `deque` in C++26, and I currently don't want to provide any way to opt-out of that behavior.

For `iterator`, the current strategy makes it store `FancyPtr<FancyPtr<T>>`. But it would make more sense to also store `FancyPtr<const FancyPtr<T>>` because we never modify the map via `iterator`.

For some pathological combinations of allocators and fancy pointers, the rebinding trick doesn't work. These cases are rejected by `static_assert`.

The existing test coverage seems to be sufficient.

frederick-vs-ja committed:
[libc++] Fix strict aliasing violation for `deque::const_iterator`

When the allocators use fancy pointers, the internal map of `deque`
stores `fancy_ptr<T>` objects, and the previous strategy accessed these
objects via `const fancy_ptr<const T>` lvalues, which usually caused
core language undefined behavior. Now `const_iterator` stores
`fancy_ptr<const fancy_ptr<T>>` instead of
`fancy_ptr<const fancy_ptr<const T>>`, and ABI break can happen when
such two types have incompatible layouts.

This is necessary for reducing undefined behavior and `constexpr`
support for `deque` in C++26, and I currently don't want to provide any
way to opt-out of that behavior.

Also removes redundant identity `static_cast` before and after type
change.

The existing test coverage seems to be sufficient.

philnik777 reviewed (changes requested):
I don't think we can do this as-is. AFAICT this changes the type of `deque::const_iterator`, which is way too big of an ABI break.

frederick-vs-ja committed:
Switch to use internally rebinding

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
`__conditional_t`

frederick-vs-ja committed:
Fix allocators and fancy pointers in `deque/abi.compile.pass.cpp`

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja commented:
@philnik777 Now I'm attempting to keep the actual type of `deque::const_iterator` unchanged and change type of the stored (possibly fancy) pointer instead.
I'm afraid that the standard seems to allow users to write member `pointer` types that can't be restored by `pointer_traits::rebind`, and thus we're still unable to handle some pathological but "correct" (standard-conforming) user-defined allocators.

frederick-vs-ja committed:
Make GCC happy

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
Make them meeting Cpp17NullablePointer

frederick-vs-ja committed:
Fix typo

frederick-vs-ja committed:
`_LIBCPP_HIDE_FROM_ABI`

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
Retarget LLVM22

frederick-vs-ja committed:
Address review comments

philnik777 commented on libcxx/docs/ReleaseNotes/22.rst:
```diff
@@ -58,5 +58,13 @@ Announcements About Future Releases
+- The ``const_iterator`` member type of ``std::deque`` is now corrected to hold a (possibly fancy) pointer to the
+  (possibly fancy) pointer allocated in the internal map. E.g. when the allocators use fancy pointers, the internal map
+  stores ``fancy_ptr<T>`` objects, and the previous strategy accessed these objects via ``const fancy_ptr<const T>``
+  lvalues, which usually caused core language undefined behavior. Now ``const_iterator`` stores
```
Ok, yes, for constant evaluation there is that difference. AFAIK the main reason for that is that some UB in the library is impossible to detect though because some UB is significantly more high level, not that there is a fundamental difference.

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja commented:
@philnik777 This seems also blocking `constexpr` addition of `stack` and `queue`. Can we merge this PR (or something like if you have another resolution) soon?

Zingam commented:
Ping

Can we proceed with this to unblock constexpr deque and subsequent work?

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

philnik777 reviewed (commented):
I think we could significantly simplify this patch by rebinding the `_MapPointer` inside `__deque_iterator` instead. I think that should remove almost all of the complexity of this patch and make it quite obvious that we don't break the symbol names.

I'd also like to see some `static_assert`s similar to the ones we added in #100585.

frederick-vs-ja committed:
Address @ldionne and @philnik777's review comments

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
Fix-up `[[__nodiscard__]]`

frederick-vs-ja committed:
Fix-up the strategy

philnik777 commented on libcxx/include/deque:
```diff
@@ -282,7 +293,18 @@ template <class _ValueType,
           >
 class __deque_iterator {
-  typedef _MapPointer __map_iterator;
+  using __map_iterator _LIBCPP_NODEBUG = __get_deque_map_iterator<_ValueType, _MapPointer>;
```
Isn't this just `__rebind_pointer_t<_MapPointer, _ValueType>`?

frederick-vs-ja replied:
It isn't. As noted in the comments, what we want is `FancyPtr<const FancyPtr<_ValueType>>` here (if the iterator is const), but not `FancyPtr<_ValueType>`.

frederick-vs-ja commented on libcxx/include/deque:
```diff
@@ -441,6 +463,11 @@ private:
+  _LIBCPP_HIDE_FROM_ABI explicit __deque_iterator(_MapPointer __m, pointer __p) _NOEXCEPT
+      : __m_iter_(__m ? pointer_traits<__map_iterator>::pointer_to(*__m) : nullptr),
+        __ptr_(__p) {}
+
```
Yeah. Looks like this just works after removing `static_cast<__map_const_pointer>()`.

frederick-vs-ja committed:
Remove unneeded constructor

philnik777 reviewed (commented):
This mostly looks good, except for the `__get_deque_map_iterator`. I think it should be possible to refactor it to something more readable.

Please also update the PR for LLVM23.

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

frederick-vs-ja committed:
Retarget LLVM 23

frederick-vs-ja committed:
Expand `__get_deque_map_iterator`

philnik777 commented on libcxx/include/deque:
```diff
@@ -283,7 +304,18 @@ template <class _ValueType,
           >
 class __deque_iterator {
-  typedef _MapPointer __map_iterator;
+  using __map_iterator _LIBCPP_NODEBUG = __get_deque_map_iterator<_ValueType, _MapPointer>;
```
Indeed, that should work. Let's change to that and I think I'm happy with this patch.

frederick-vs-ja committed:
Address review comments

- Also change `deque::iterator`.
- Revert changes in the test.
- Update Release notes.

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator

# Conflicts:
#	libcxx/docs/ReleaseNotes/23.rst

frederick-vs-ja committed:
Fix pointer rebinding

frederick-vs-ja committed:
Simplify member type alias and remove unnecessary casts

frederick-vs-ja committed:
Merge branch 'main' into fix-deque-const_iterator
